Pyfunc: Simplify process for adding custom inference logic and data/object dependencies

dbczumar opened the pull request:
This PR introduces aims to make it easier for users to incorporate custom inference logic and data dependencies into an MLflow model, as well as compose multiple models together. Changes primarily impact the `mlflow.pyfunc` APIs.

When reviewing, please build and examine the documentation. While the diff looks large, more than two-thirds of it consists of tests (including a rewrite of `tests/pyfunc/test_model_export` to use the `pytest` format instead of `unittest`) and documentation. The majority of changes can be found in `mlflow/pyfunc/model.py`.

The following is some sample code followed by the MLmodel configuration and directory structure that it produces:

**Sample code (run in an interpreter or __main__)**

```
import sklearn.datasets
import sklearn.neighbors

import mlflow
import mlflow.pyfunc
import mlflow.sklearn

iris = sklearn.datasets.load_iris()
x = iris.data[:, :2]
y = iris.target
sklearn_knn_model = sklearn.neighbors.KNeighborsClassifier()
sklearn_knn_model.fit(x, y)

class CustomSklearnModel(mlflow.pyfunc.PythonModel):

    def __init__(self, predict_fn):
        super(CustomSklearnModel, self).__init__()
        self.predict_fn = predict_fn

    def load_context(self, context):
        self.model = mlflow.sklearn.load_model(path=context.artifacts["sk_model"])

    def predict(self, model_input):
        return self.predict_fn(self.model, model_input)

def predict_times_two(model, model_input):
    return model.predict(model_input) * 2

sklearn_artifact_path = "sk_model_artifact"
with mlflow.start_run():
    mlflow.sklearn.log_model(sk_model=sklearn_knn_model, artifact_path=sklearn_artifact_path)
    sk_model_artifact_uri = mlflow.get_artifact_uri(sklearn_artifact_path)

with mlflow.start_run():
    my_model = CustomSklearnModel(predict_times_two)
    mlflow.pyfunc.log_model(artifact_path="pyfunc_model",
                            artifacts={
                                "sk_model": sk_model_artifact_uri
                            },
                            python_model=my_model)
```

**Produced model config**

```
artifact_path: pyfunc_model
flavors:
  python_function:
    artifacts:
      sk_model:
        path: artifacts/sk_model_artifact
        uri: /private/tmp/mlruns/0/7f8bce9c28ef43f1b8ec92ce7f2889ec/artifacts/sk_model_artifact
    env: conda.yaml
    loader_module: mlflow.pyfunc.model
    python_model: model_class.pkl
    python_version: 2.7.14
run_id: 523f2374d9134d67860f57d8d9a308c3
utc_time_created: '2019-01-08 00:51:45.248224'
```

**Produced directory structure**
```
pyfunc_model/
├── MLmodel
├── artifacts
│   └── sk_model_artifact
│       ├── MLmodel
│       ├── conda.yaml
│       └── model.pkl
├── conda.yaml
├── model_class.pkl
```

dbczumar committed:
Pyfunc progress

dbczumar committed:
Model changes

dbczumar committed:
Artifact resolution

dbczumar committed:
Validation

dbczumar committed:
Artifact validation, handle class name imports

dbczumar committed:
Add a few tests, many to come

dbczumar committed:
Add backticks to warning / exception text for readability

dbczumar committed:
Artifact test

dbczumar committed:
Fix relative path resolution in azure blob artifact repo

dbczumar committed:
Add test

dbczumar committed:
Naming change

dbczumar committed:
Merge azure fix and master

dbczumar committed:
Reorder tests

dbczumar committed:
Test updates

dbczumar committed:
Validate that blob names begin with artifact path

dbczumar committed:
Add negative test case

dbczumar committed:
Lint

dbczumar committed:
azure test fix

dbczumar committed:
DBFS test progress

dbczumar committed:
Implement artifact uri convenience methods with preliminary azure test

dbczumar committed:
Artifact repos define path module

dbczumar committed:
Use path module for _download_artifact_from_uri

dbczumar committed:
Address comments

dbczumar committed:
More comments

dbczumar committed:
FTP test fix

dbczumar committed:
Test fixes

dbczumar committed:
Tests, get_artifact_uri changes

dbczumar committed:
Lint

dbczumar committed:
Import fix

dbczumar committed:
Lint

dbczumar committed:
Test output directory download

dbczumar committed:
Lint

dbczumar committed:
Merge artifact uri resolution changes

dbczumar committed:
Tests

dbczumar committed:
Fix merge conflicts

dbczumar committed:
Merge branch 'artifact_uri_methods' into pyfunc_v2

dbczumar committed:
Docs

dbczumar committed:
Add kwargs parameter to pyfunc 'add_to_model'

dbczumar committed:
Merge pyfunc kwargs change

dbczumar committed:
Import logging

dbczumar committed:
Add pyfunc utils module

dbczumar committed:
Fix cli import

dbczumar committed:
Lint

dbczumar committed:
Docs

dbczumar committed:
Merge pyfunc utils changes

dbczumar committed:
Import fix

dbczumar committed:
Revert doc change

dbczumar committed:
Merge branch 'pyfunc_module_clean' into pyfunc_v2

dbczumar committed:
Docs changes

dbczumar committed:
Revert dbfs test changes

dbczumar committed:
Revert init changes

dbczumar committed:
Docs improvements

dbczumar committed:
More docs

dbczumar committed:
Test fixes and improvements

dbczumar committed:
Address comments - remove utils module

dbczumar committed:
Use isdir

dbczumar committed:
Path

dbczumar committed:
Remove utils module

dbczumar committed:
Merge utils changes

dbczumar committed:
Merge master

dbczumar committed:
Improve test naming

dbczumar committed:
Conda env test fixture

dbczumar committed:
Conda tests

dbczumar committed:
Make artifacts and params optional, test this use case

dbczumar committed:
Merge master

dbczumar committed:
Small docs tweaks

dbczumar committed:
Revert changes to tracking init file

dbczumar committed:
Revert azure blob tests

dbczumar committed:
Add docker test

dbczumar committed:
Change temporary directory behavior

dbczumar committed:
Docs fixes

dbczumar committed:
Remove comma

dbczumar committed:
Lint

dbczumar committed:
Import reversion

dbczumar committed:
Docs comments

dbczumar committed:
Fix comment syntax

dbczumar committed:
Capitalization

dbczumar committed:
Fix path naming duplication issue

dbczumar committed:
Fix version doc

dbczumar committed:
Fix lint errors and remove redundant test

dbczumar committed:
Fix spark UDF tests

dbczumar committed:
Import fixes

dbczumar commented on tests/sagemaker/test_model_export.py:
```diff
@@ -1,71 +0,0 @@
-from __future__ import print_function
```
This suite consists of a single test that attempts to score a pyfunc-wrapped scikit-learn model by deploying it inside a Docker container. A similar test has been implemented in the `tests.pyfunc.test_model_export` suite; it is called `test_sagemaker_docker_model_scoring_with_default_conda_env`. Therefore, we propose removing this test suite.

mparkhe commented on mlflow/pyfunc/__init__.py:
```diff
@@ -141,6 +143,127 @@ def add_to_model(model, loader_module, data=None, code=None, env=None, **kwargs)
     return model.add_flavor(FLAVOR_NAME, **parms)
 
 
+def save_model(path, model_class, artifacts=None, parameters=None, conda_env=None, code_paths=None,
```
There is a change in signature. That would be a problem.

dbczumar replied:
I've implemented a backwards-compatible API. Please take a look!

mparkhe commented on mlflow/pyfunc/__init__.py:
```diff
@@ -318,106 +441,3 @@ def predict(*args):
     return pandas_udf(predict, result_type)
-
-
-def save_model(dst_path, loader_module, data_path=None, code_path=(), conda_env=None,
```
Please bring `save_model` and `load_model` back down here so it will be easy to compare difference.

dbczumar replied:
I've moved `save_model` and `log_model` to `mlflow.pyfunc.model`. The logical flow differs substantially from the old method. It seems easier to review it as a completely new implementation rather than a modification of existing conventions. Please let me know if you disagree.

mparkhe replied:
Nope this is cool!
The design doc and inline documentation makes it super easy to follow. Thank you.

mparkhe commented on mlflow/pyfunc/__init__.py:
```diff
@@ -89,9 +89,11 @@
 from mlflow.tracking.fluent import active_run, log_artifacts
 from mlflow import tracking
 from mlflow.models import Model
+from mlflow.pyfunc.model import _save_model, PythonModel, PythonModelContext
```
Is there a reason `_save_model` is in `model.py` if the only caller is in this file. If you want to hide it -- move `save_model` to another file and then alias it here

dbczumar replied:
Good call, the reason was to produce documentation for `save_model` under the `mlflow.pyfunc` module name, rather than `mlflow.pyfunc.model`. I've instead renamed `mlflow.pyfunc.model._save_model` to `mlflow.pyfunc.model.save_model` and moved `log_model` to the `mlflow.pyfunc.model` module. I've updated the sphinx configuration to document these functions under the `mlflow.pyfunc` namespace, since they are imported.

mparkhe commented on mlflow/pyfunc/model.py:
```diff
@@ -0,0 +1,333 @@
+            cloudpickle.dump(model_class, out)
+        saved_model_class_config[CONFIG_KEY_MODEL_CLASS_PATH] = saved_model_class_subpath
+    else:
+        saved_model_class_config[CONFIG_KEY_MODEL_CLASS_NAME] = model_class
```
this sounds a little confusing: `inspect.isclass(model_class) == False` here so not clear why `model_class` is saved as `CONFIG_KEY_MODEL_CLASS_NAME`

dbczumar replied:
Agreed! I've added comments during the model class serialization and deserialization procedures to explain the logical flow. In this case, `inspect.isclass(model_class) == False` implies that `model_class` is not a Python class object. Therefore, we assume that it's the fully-qualified name of a Python class and save it using the key `CONFIG_KEY_MODEL_CLASS_NAME`.

dbczumar replied:
Update: I've removed the string-based definition of `model_class`. Now, it must always be a Python class object. This removes the confusion you identified.

mparkhe reviewed (changes requested):
Some changes would be required to accommodate existing clients. Also recommending some changes to ensure that diffs are minimized.

dbczumar committed:
Document imported functions from mlflow.pyfunc.model, rather than wrapping a _save_model method

tomasatdatabricks reviewed (commented):
Why thing I don't understand is, are you trying to replace current pyfunc with this? I thought this would be and add on flavor. This way we would loose backwards compatibility and all the flavors would need to be modified?

dbczumar commented on mlflow/pyfunc/model.py:
```diff
@@ -0,0 +1,421 @@
+                       instances of ``model_class``.
+    :param mlflow_model: The model configuration to which to add the ``mlflow.pyfunc`` flavor.
+    """
+    return Model.log(artifact_path=artifact_path, flavor=mlflow.pyfunc, artifacts=artifacts,
```
I'll change it to `mlflow.pyfunc.model`, but this works because `mlflow.pyfunc` imports `save_model` from `mlflow.pyfunc.model`. there are several tests for this use case.

tomasatdatabricks replied:
Ah I missed that. No need to change int in that case!

dbczumar committed:
Add conda to test requirements

dbczumar committed:
readable MatchSpec import error case with helpful explanation

dbczumar committed:
Improve error message

dbczumar committed:
Add explanatory comments for model class serialization/deserialization behavior

dbczumar committed:
Fix indentation

dbczumar committed:
Lint

dbczumar commented:
In response to review comments, I've written the following document outlining the pre-existing use case satisfied by the `mlflow.pyfunc.save_model` API, the use case satisfied by the new `save_model` API, and a proposal for merging the APIs to satisfy both use cases. Please leave feedback.

# Backwards compatibility: Use cases and proposed changes to MLflow’s custom pyfunc APIs

In this short doc, we perform the following tasks:

1. First, we detail the [existing workflow for custom pyfunc serialization](#existing-custom-pyfunc-serialization-workflow) and the API that supports it. We refer to the supporting MLflow API as `save_model_1()`.


2. Then, we detail an [additional workflow](#additional-pyfunc-serialization-workflow) that is supported by [this PR](https://github.com/mlflow/mlflow/pull/793). We refer to the supporting MLflow API as `save_model_2()`.


3. We [propose a solution](#a-combined-api-for-both-workflows) for combining `save_model_1()` and `save_model_2()` to produce a single `save_model()` API.

## Existing custom Pyfunc serialization workflow

The current `mlflow.pyfunc.save_model/log_model` APIs satisfy the following use case (for an example, see https://github.com/mlflow/mlflow/blob/18d5abf7c0dc7e023d246adc1130f112366f5622/examples/flower_classifier/image_pyfunc.py#L34-L170. The `log_model()` method invokes Steps 1 and 3):

1. The user collects / has collected some files (serialized models, CSV’s, etc) in a directory on their local filesystem. The path to this directory is called the **data path**.

2. The user writes / has written some code that interprets the contents located at **data path** as an MLflow model. This code contains the following method definition:

   ```py
   def _load_pyfunc(*data_path*):
     """
     :return: A class that defines a method with the following 
              signature: predict(model_input)
     """
   ```

   This code is referenced as a collection of paths, called **code_paths**. The name of the Python module containing the definition of `_load_pyfunc()` is called the **loader_module**.

3. The user saves the model in MLflow Model format with the pyfunc flavor by invoking the following MLflow API:

   ```py
   def save_model_1(*data_path*, loader_module, code_paths, 
                    conda_env, model=mlflow.Model()):
    """
    :param *data_path*: The path to the directory containing
                        User-collected files.
    :param *loader_module*: The name of the Python module 
                            containing the definition of 
                            ``_load_pyfunc()``
    :param *code_paths*: A collection of paths to Python files,
                       including the file that defines the 
                       **loader_module**.
    :param conda_env: A conda environment to activate before
                      loading the model.
    :param model: The MLModel configuration to which to add
                  the "pyfunc" flavor.
    """
   ```

## Additional Pyfunc serialization workflow

[This PR](https://github.com/mlflow/mlflow/pull/793) introduces APIs that satisfy a different use case. It includes an example workflow under the Sample Code header. Procedure:

1. The user selects some file dependencies that may exist on the local filesystem or remotely (on S3, Azure, etc) and assigns them names by producing an **artifacts** Python dictionary. For example,

   ```py
    *artifacts* = {

        "sklearn_model": "s3://<bucket_name>/path/to/sklearn/model,
        "als_features": "/local/path/to/serialized/features",
        ...

    }
   ```

2. The user selects some Python object dependencies and assigns them names by producing a **parameters** dictionary.

   ```py
   *parameters* = {

      "matrix": np.random.rand(10,20),
      "user_names": ["Thad", "Tad", "Chad"]

   }
   ```

3. The user writes / has written a model class that uses the specified **artifacts** and **parameters** to perform inference. The class has the following format:

   ```py
   class CustomModel(mlflow.pyfunc.PythonModel):

        def __init__(self, context):
            """
            :param context: A bundle containing user-specified
                            *artifacts* and *parameters*.
            """
            # Load artifacts from the filesystem and assign them
            # to class attributes here

        def predict(self, model_input):
            # Perform inference
   ```

4. The user creates an MLflow model with the pyfunc flavor by invoking the following API:

   ```
   def save_model_2(*artifacts*, *parameters*, *model_class*, 
                    *code_paths*, conda_env, model=mlflow.Model()):
        """
        :param *artifacts*: A dictionary mapping user-specified file 
                            / artifact names to their URI’s or
                            filesystem paths
        :param *parameters*: A dictionary mapping user-specified
                             names to Python objects
        :param *model_class*: A Python class object (of type 
                              ``type``), as defined in Step 3.
        :param *code_paths*: A collection of paths to Python files
                             that *model_class* may depend on. If 
                             *model_class* was imported from another 
                             module, this should include the module 
                             that defines *model_class*.
        :param conda_env: A conda environment to activate before
                          loading the model.
        :param model: The MLModel configuration to which to add
                      the "pyfunc" flavor.
        """
   ```

## A combined API for both workflows

We propose defining a `save_model()` API that is the union of `save_model_1()` and `save_model_2()` with the following signature:

```
def save_model(data_path=None, loader_module=None, code_paths=None, 
               conda_env=None, model=mlflow.Model(),
               artifacts=None, parameters=None, model_class=None)
```

Note that all parameters have default values. We propose that users must specify *either* the **first row of parameters** or the **third row of parameters** exclusively. The second row can be specified for both use cases. We will extensively document the use case / workflow that each set of parameters is meant to support.

dbczumar committed:
Remove conda from test requirements

dbczumar committed:
Add usage: conda [-h] [-V] command ...

conda is a tool for managing and deploying applications, environments and packages.

Options:

positional arguments:
  command
    clean        Remove unused packages and caches.
    config       Modify configuration values in .condarc. This is modeled
                 after the git config command. Writes to the user .condarc
                 file (/Users/czumar/.condarc) by default.
    create       Create a new conda environment from a list of specified
                 packages.
    help         Displays a list of available conda commands and their help
                 strings.
    info         Display information about current conda install.
    install      Installs a list of packages into a specified conda
                 environment.
    list         List linked packages in a conda environment.
    package      Low-level conda package utility. (EXPERIMENTAL)
    remove       Remove a list of packages from a specified conda environment.
    uninstall    Alias for conda remove. See conda remove --help.
    search       Search for packages and display associated information. The
                 input is a MatchSpec, a query language for conda packages.
                 See examples below.
    update       Updates conda packages to the latest compatible version. This
                 command accepts a list of package names and updates them to
                 the latest versions that are compatible with all other
                 packages in the environment. Conda attempts to install the
                 newest versions of the requested packages. To accomplish
                 this, it may update some packages that are already installed,
                 or install additional packages. To prevent existing packages
                 from updating, use the --no-update-deps option. This may
                 force conda to install older versions of the requested
                 packages, and it does not prevent additional dependency
                 packages from being installed. If you wish to skip dependency
                 checking altogether, use the '--force' option. This may
                 result in an environment with incompatible packages, so this
                 option must be used with great caution.
    upgrade      Alias for conda update. See conda update --help.

optional arguments:
  -h, --help     Show this help message and exit.
  -V, --version  Show the conda version number and exit.

conda commands available from other packages:
  build
  convert
  develop
  env
  index
  inspect
  metapackage
  render
  server
  skeleton
  verify package to test environment in Travis

dbczumar committed:
Backwards compatibility part 1

dbczumar committed:
More backwards compat

dbczumar committed:
More backwards compat and testing

dbczumar committed:
Lots of documentation improvements

dbczumar committed:
Param renaming

dbczumar committed:
Docs spacing, lint

dbczumar commented:
@mparkhe @tomasatdatabricks I've addressed your comments regarding backwards compatibility and introduced a significant amount of documentation about the workflows for creating custom pyfunc models. Please take another pass!

dbczumar committed:
Revert line addition

dbczumar committed:
Docs tweaks for 'models' section

dbczumar committed:
Test cleanup

dbczumar committed:
Test fixes

tomasatdatabricks commented on mlflow/pyfunc/__init__.py:
```diff
@@ -152,6 +293,244 @@ def _load_model_env(path, run_id=None):
     return _get_flavor_configuration(model_path=path, flavor_name=FLAVOR_NAME).get(ENV, None)
 
 
+def save_model(dst_path, loader_module=None, data_path=None, code_path=None, conda_env=None,
```
Personally, I would prefer adding the new stuff as a separate flavor or PyFuncBuilder. Or move the older api to somewhere as a lower level api to construct pyfunc. I think it is really confusing to have one function with 2 conflicting sets of parameters.

dbczumar replied:
I also agree that this may be somewhat confusing. However, after discussing it with @mparkhe and @mateiz, we agree that this seems more pythonic than defining multiple `save_model` methods.

tomasatdatabricks commented on mlflow/pyfunc/__init__.py:
```diff
@@ -1,19 +1,32 @@
-
-Pyfunc model format is defined as a directory structure containing all required data, code, and
+The ``mlflow.pyfunc`` module defines a generic :ref:`filesystem format <pyfunc-filesystem-format>`
+for Python models and provides utilities for saving to and loading from this format. The format is
```
That's really good documentation but I am not sure if it's not too much for in-code docs. I wonder if large bulk of it belongs to the docs instead? 

In my opinion, in-code doc should give functional description necessary to work with the code. E.g. the paragraph starting ```Many of MLflow's model persistence modules ,...``` sounds to me like something that might belong to docs instead of in code. What do you think?

dbczumar replied:
This documentation gets rendered and displayed in the "Python API" section of the MLflow docs website. It also provides a reference for how to use the `save_model()` and `log_model()` APIs, so it seems relevant to the code in that sense.

I agree that the `Many of MLflow's model persistence...` portion is probably a better fit for the "Models" documentation section. I've moved it there instead.

dbczumar committed:
Test fixes

dbczumar committed:
Diff fix

tomasatdatabricks reviewed (approved):
lgtm

I have few comments (opinions) but I am ok if you decide to leave it as is:

a) I am not sure if all the new in-code documentation belongs to in-code or large part of it should be moved to docs instead. Seems a bit chatty for in-code docs, at least to me.

b) Personally I don't like having an api with two incompatible modes of operation. I would prefer to rename and move the new or old way of generating pyfunc rather than having one function with two sets of parameters.

dbczumar committed:
Diff reversions

dbczumar committed:
Spark test load_pyfunc signature change

dbczumar committed:
Move some docs from code to 'models'

dbczumar committed:
Test for argument incompatibilities

dbczumar committed:
Loader module test fix

dbczumar committed:
Move _copy_file_or_tree tests from pyfunc utils to file utils

mparkhe commented on docs/source/models.rst:
```diff
@@ -113,80 +113,28 @@ for saving and loading models to and from this format. The format is self-contai
 .. code:: python
 
-    predict(data: pandas.DataFrame) -> [pandas.DataFrame | numpy.array]
+    predict(data: pandas.DataFrame) -> numpy.ndarray | pandas.Series | pandas.DataFrame
```
I think you need the box brackets here, to suggest one of.

```
    predict(data: pandas.DataFrame) -> [numpy.ndarray | pandas.Series | pandas.DataFrame]
```

Looks like this on [mlflow.org/.../models.html](https://mlflow.org/docs/latest/models.html) documentation

![image](https://user-images.githubusercontent.com/6900999/51003310-0c6f4680-14eb-11e9-98b9-3762d7b1e956.png)

dbczumar replied:
Good catch. I've added brackets to fix the documentation.

mparkhe commented on mlflow/pyfunc/__init__.py:
```diff
@@ -320,62 +452,246 @@ def predict(*args):
+        return mlflow.pyfunc.model._save_model_with_class_artifacts_params(
+            path=dst_path, model_class=model_class, artifacts=artifacts, parameters=parameters,
+            conda_env=conda_env, code_paths=code_path, mlflow_model=model)
+    elif data_path is not None:
```
Does it make sense to these error checks before execution on `save_model_1` vs `save_model_2` recipe...

dbczumar replied:
Definitely! I've modified the logic to perform parameter validation in each internal method.

mparkhe commented on mlflow/pyfunc/model.py:
```diff
@@ -0,0 +1,367 @@
+        return self._parameters
+
+
+def _save_model_with_class_artifacts_params(path, model_class, artifacts=None, parameters=None,
```
👍 on how you broke this down into class vs loader based and forking it into individual flows! Awesome!

dbczumar replied:
Thanks Mani!

mparkhe reviewed (approved):
This is a large diff -- but thanks for putting together a design doc that describes the change and the detailed inlined documentation and tests. This is a great feature!

There are some broken tests -- look at Travis logs. But otherwise LGTM!

dbczumar committed:
Test fixes

dbczumar committed:
Docs inference API adjustment

dbczumar committed:
Validate parameters in each internal _save method

dbczumar committed:
Test fixes

dbczumar committed:
Lint

dbczumar committed:
Fix test patching

aarondav commented on mlflow/pyfunc/model.py:
```diff
@@ -0,0 +1,379 @@
+                      ``artifact_path`` is the absolute filesystem path to the artifact.
+    """
+    try:
+        from conda.resolve import MatchSpec
```
This seems relatively complicated to require a conda dependency (discoverable only at runtime) to do some conda-specific logic about resolving cloudpickle versions. What happens if cloudpickle is a transitive dependency?

Could we just print this warning at load time, as opposed to save time, like we do python version? Anyway, the runtime environment may not precisely match what we write (e.g., spark_udf does not currently use the conda env), so this is the only place we can be relatively sure, and anything else is just a complicated best guess.

aarondav replied:
Whether or not we end up changing how we do this, I think it might be a good idea to move the validation logic and any associated tests to a follow-up PR, where we can discuss the complexities independently. Otherwise, it's a relatively minor aspect of this feature, but I think it contributes significantly to the complexity of the PR.

Also, a second point is that if the end goal of this logic is to print best-effort warnings, probably a failure to import conda should not be an exception, but just a warning that we could not do our validation.

dbczumar replied:
I agree that we shouldn't require a conda dependency for this validation logic. However, if the `conda.resolve` dependency happens to exist, I think it's reasonable to perform a check for conflicting library versions at model save time. If cloudpickle is a transitive dependency, it just means that we won't capture potentially conflicting state.

Checking for version compatibility at model load time also seems like a good idea. 

Keeping with your suggestion, I've removed the validation and associated tests from this PR; I'll file a separate one with some changes to the validation logic that avoids a strict dependency on `conda.resolve` and validates during model load.

aarondav commented:
Are parameters always functions? Not sure I understand how the serialized form can be `predict_fn: parameters/predict_fn.pkl`, that seems like it would be equivalent if I wanted predict_fn to be a string.

aarondav commented on docs/source/models.rst:
```diff
@@ -113,80 +113,28 @@ for saving and loading models to and from this format. The format is self-contai
 that it includes all the information necessary to load and use a model. Dependencies
 are stored either directly with the model or referenced via Conda environment.
 
+Many of MLflow's model persistence modules, such as :mod:`mlflow.sklearn`, :mod:`mlflow.keras`,
```
Would be good to get @stbof to review these doc changes.

aarondav commented on mlflow/pyfunc/__init__.py:
```diff
@@ -78,6 +87,125 @@
+           - ``predict()``: This method leverages **artifacts** and **parameters** to evaluate
+             queries. It must adhere to the :ref:`pyfunc-inference-api`.
+
+       The **model_class** must inherit from :class:`~PythonModel`. It is specified via the
```
Maybe we could just say "In this case, you must provide a Python class which inherits from :class:`~PythonModel`, with two methods:" and remove this paragraph.

stbof replied:
In this case, you must provide a Python class that inherits from :class:`~PythonModel`, with two methods:

dbczumar replied:
Removed the paragraph and changed the text to read:

```
       In this case, you must provide a Python class which inherits from :class:`~PythonModel`
       and defines two methods: ``__init__()`` and ``predict()`. This class is specified via the
       ``model_class`` parameter and is automatically serialized and deserialized as a Python
       class object.
```

aarondav commented on mlflow/pyfunc/__init__.py:
```diff
@@ -78,6 +87,125 @@
+           - ``__init__()``: This method loads and interprets **artifacts** from the MLflow
+             model directory.
+
+             *Note:* **parameters** are *automatically* read from the MLflow model directory as
```
This is a lot of italics and bold. Maybe just remove the italics?

dbczumar replied:
Done! Removed the italics.

aarondav commented on mlflow/pyfunc/__init__.py:
```diff
@@ -78,6 +87,125 @@
+
+       If you already have a collection of model data, :meth:`save_model()` and :meth:`log_model()`
+       can import the data as an MLflow model. Model data may consist of both **artifacts** and
+       *serialized* **parameters**. The ``data_path`` argument specifies the local filesystem path
```
I wonder if instead of `parameters`, another name like `data` or `model_data` would be more obvious. If I conceptualize something as a file (e.g., customers.csv), I would probably pass in the name "customers.csv" to parameters, as opposed to the result of `open("customers.csv").read()`.

dbczumar replied:
After offline discussion, I've removed `parameters` and replaced `model_class` with `python_model`, where the latter term now refers to a python class **instance**, rather than a class **type**. This instance is now serialized and deserialized.

stbof commented on docs/source/models.rst:
```diff
@@ -113,80 +113,28 @@ for saving and loading models to and from this format. The format is self-contai
 that it includes all the information necessary to load and use a model. Dependencies
 are stored either directly with the model or referenced via Conda environment.
 
+Many of MLflow's model persistence modules, such as :mod:`mlflow.sklearn`, :mod:`mlflow.keras`,
```
Many MLflow model persistence modules

dbczumar replied:
Done - thanks!

stbof commented on docs/source/models.rst:
```diff
@@ -113,80 +113,28 @@ for saving and loading models to and from this format. The format is self-contai
+containing the ``pyfunc`` flavor, regardless of which persistence module or framework was used to
+produce the model. *This interoperability is very powerful because it allows any Python model to be
+productionized in a variety of environments*.
+
```
Remove the italics.

dbczumar replied:
Italics removed!

dbczumar committed:
Remove validation, to be submitted in a separate PR

dbczumar committed:
Address docs comments

dbczumar committed:
Remove 'conda' dependency from travis tests

dbczumar committed:
Remove , require a class instance rather than type obj

dbczumar committed:
Docs

dbczumar committed:
Docs tweaks

dbczumar committed:
Docs

dbczumar committed:
More docs tweaks

dbczumar committed:
Import fixes

dbczumar committed:
Test fixes

aarondav commented on mlflow/pyfunc/__init__.py:
```diff
@@ -320,62 +433,192 @@ def predict(*args):
+                            ]
+                        }
+
+    :param python_model: An instance of a subclass of :class:`~PythonModel`. This class will be
```
for some reason this looks weird (py3 make):

![image](https://user-images.githubusercontent.com/1400247/51126064-0ae99b00-17d7-11e9-8068-5988b123b721.png)

aarondav commented on mlflow/pyfunc/__init__.py:
```diff
@@ -320,62 +433,192 @@ def predict(*args):
+        "data_path": data_path,
+    }
+    second_argument_set = {
+        "artifacts": artifacts,
```
Empirically, why could `artifacts` not be specified if I'm using a loader module? I understand the desire for either a tuple of (python file, entrypoint) vs python instance, since those map to how we construct the python object representation of the model, but artifacts seem orthogonal.

I wonder if during the MLflow 1.0 work we should consider revamping the naming of this API a bit -- `data_path` and `loader_module` do not seem related, but they both need to be specified together or not at all.

aarondav commented on mlflow/pyfunc/__init__.py:
```diff
@@ -320,62 +433,192 @@ def predict(*args):
+    }
+    second_argument_set = {
+        "artifacts": artifacts,
+        "python_model": python_model,
```
Btw, for my understanding, would we expect imports to be captured by the serialization of python_model? I would guess not, but this does imply that if I have more than one python file, I may actually still want to use something like `data_path`, right?

aarondav commented on mlflow/pyfunc/__init__.py:
```diff
@@ -320,62 +433,192 @@ def predict(*args):
+                    first_set_entries=first_argument_set,
+                    second_set_entries=second_argument_set)),
+            error_code=INVALID_PARAMETER_VALUE)
+    elif not (first_argument_set_specified or second_argument_set_specified):
```
Is this logically sound? I seems like I could specify only `artifacts` and not get any error.

aarondav commented on mlflow/pyfunc/model.py:
```diff
@@ -0,0 +1,275 @@
+import os
```
Mind adding a file-level doc of what we expect to find here?

aarondav commented on mlflow/pyfunc/model.py:
```diff
@@ -0,0 +1,275 @@
+from mlflow.utils.model_utils import _get_flavor_configuration
+from mlflow.utils.file_utils import TempDir, _copy_file_or_tree
+
+DEFAULT_CONDA_ENV = _mlflow_conda_env(
```
This is only the default for serialized PythonModel objects. I guess this file is only concerned with those, although from a logical perspective, I would imagine the code to be shared almost identically with the data_path/loader_module pathway. For example, we could imagine just instantiating a default PythonModel which loads the provided loader_module -- I guess the main downside of that approach is that we would need to require cloudpickle for all pyfunc model environments, which may actually be pretty severe.

Anyway, all that to say, we should at least comment at this point in the code what this is the default for, in addition to having the module-level documentation saying that all concepts here are related only to PythonModel serialization (as opposed to loader_module saving of pyfunc models; unless I'm wrong about that distinction).

aarondav replied:
I wonder if we should consider at some point shading cloudpickle into mlflow, to fix the version that we depend on and make it independent of the customer's environment.

aarondav commented on mlflow/pyfunc/model.py:
```diff
@@ -0,0 +1,275 @@
+                        can use to perform inference.
+        """
+        # pylint: disable=attribute-defined-outside-init
+        self.context = context
```
Hmm. I'm not sure this makes sense -- it is quite reasonable to forget to call the parent load_context method, in which case a contract such as "self.context is available" would be violated. Maybe simpler would be to pass context into `predict`, such that `load_context` becomes merely a hook for optimization.

aarondav commented on mlflow/pyfunc/model.py:
```diff
@@ -0,0 +1,275 @@
+        :param context: A :class:`~PythonModelContext` instance containing artifacts that the model
+                        can use to perform inference.
+        """
+        # pylint: disable=attribute-defined-outside-init
```
haha, kind of funny that this is un-pythonic.

aarondav commented on mlflow/pyfunc/model.py:
```diff
@@ -0,0 +1,275 @@
+                            exist for as long as the :class:`~PythonModelContext` is in scope.
+        """
+        self._artifacts = artifacts
+        self._directory_managers = directory_managers
```
What is this and what is the non-existent `directories` parameter?

aarondav commented on mlflow/pyfunc/model.py:
```diff
@@ -0,0 +1,275 @@
+        """
+        :param artifacts: A dictionary of ``<name, artifact_path>`` entries, where ``artifact_path``
+                          is an absolute filesystem path to a given artifact.
+        :param directories: A list of objects managing the lifecycle of directories that the
```
I guess this should be `directory_managers`.

aarondav commented on mlflow/pyfunc/model.py:
```diff
@@ -0,0 +1,275 @@
+                             python_model_type=type(python_model))),
+                error_code=INVALID_PARAMETER_VALUE)
+
+    if artifacts is not None and len(artifacts) > 0:
```
Maybe just `if artifacts:`?

aarondav commented on mlflow/pyfunc/model.py:
```diff
@@ -0,0 +1,275 @@
+        # Because `tempfile.TemporaryDirectory` does not exist prior to Python 3.2, create an
+        # unmanaged temporary directory instead. Depending on the system, this directory is likely
+        # to be created in "/var" or "/tmp" and will be removed on system reboot.
+        # TODO: If the longevity of the temporary directory prior to Python 3.2 becomes problematic,
```
This is probably a premature optimization in general; would rather add this directory_managers feature later.

aarondav commented on tests/pyfunc/test_spark.py:
```diff
@@ -33,44 +32,78 @@ def score_model_as_udf(model_path, run_id, pandas_df, result_type="double"):
-                             dtype=types[i]) for i in range(len(prediction))},
-            columns=[str(i) for i in range(len(prediction))])
-        return prediction_df
+def get_model_class():
```
Why is this in a method?

aarondav commented on tests/pyfunc/test_spark.py:
```diff
@@ -33,44 +32,78 @@ def score_model_as_udf(model_path, run_id, pandas_df, result_type="double"):
+@pytest.mark.large
+def test_spark_udf(spark, main_scoped_model_class, tmpdir):
+    class_instance_model_path = os.path.join(str(tmpdir), "class_model")
+    mlflow.pyfunc.save_model(
```
Maybe comment at the top of this test that we plan to save the model in two different ways (although ideally these are two separate tests which just share code; if this test fails for one of the two branches, it may be challenging to figure out which one failed).

aarondav replied:
I also don't love that you felt the need to add an integration test here; ideally, the contract and architecture of pyfunc save_model should ensure that it's not possible that one works and the other doesn't. This might be an argument for implementing save_model with loader_module in terms of PythonModel.

dbczumar committed:
Address review comments

dbczumar committed:
Add custom pyfunc example

dbczumar committed:
Revert "Add custom pyfunc example"

This reverts commit 791359087a77b6e2a105f700aed7b684e98d83d9.

dbczumar committed:
Rename user-defined predict() to _predict() and define predict() at load time as a lambda

dbczumar committed:
Revert "Rename user-defined predict() to _predict() and define predict() at load time as a lambda"

This reverts commit 8708d1b63be7899d5226197fa1b6d3af2ef52aab.

dbczumar committed:
Fix PythonModel evaluation by introducing a PyfuncWrapper class and update docs

dbczumar committed:
Lint

dbczumar reviewed (commented):
@aarondav Addressed your comments!

dbczumar committed:
Revert linter changes

dbczumar commented on mlflow/pyfunc/__init__.py:
```diff
@@ -320,62 +433,192 @@ def predict(*args):
-    Export model as a generic Python function model.
+    Create a custom Pyfunc model, incorporating custom inference logic and data dependencies.
+
+    For information about the workflows that this method supports, please see :ref:`"workflows for
```
Filed https://github.com/mlflow/mlflow/pull/807 to address this

aarondav commented on mlflow/pyfunc/model.py:
```diff
@@ -270,6 +205,21 @@ def _load_pyfunc(model_path):
+    return _PythonModelPyfuncWrapper(python_model=python_model, context=context)
+
+
+class _PythonModelPyfuncWrapper:
```
Add a class doc saying which interface this is extending

dbczumar replied:
Added the same class doc in used by other wrappers (ex: the one in the `mlflow.pytorch` module)

aarondav commented on tests/pyfunc/test_model_export_with_class_and_artifacts.py:
```diff
@@ -57,7 +57,6 @@ class ModuleScopedSklearnModel(get_model_class()):
     """
     A custom Python model class defined in the test module scope.
     """
-    pass
```
Interesting

dbczumar replied:
Turns out the `pass` statement is unnecessary

dbczumar committed:
Add class doc
